dissect/target/plugins/os/windows/sam.py

```diff
@@ -137,83 +210,225 @@
+
+        r = b""
+        for key in syskey_keys:
+            r += bytes.fromhex(lsa.subkey(key).class_name)
+
+        return bytes(r[i] for i in permutation_matrix)
+
+    def calculate_samkey(self, syskey: bytes) -> bytes:
+        aqwerty = b"!@#$%^&*()qwertyUIOPAzxcvbnmQQQQQQQQQQQQ)(*@&%\0"
+        anum = b"0123456789012345678901234567890123456789\0"
+
+        f_reg = self.target.registry.key(self.SAM_KEY).value("F").value
+        f = c_sam.DOMAIN_ACCOUNT_F(f_reg)
+        f_key = f_reg[len(c_sam.DOMAIN_ACCOUNT_F) :]
+        fk = c_sam.SAM_KEY(f_key)
+
+        if f.revision not in [0x02, 0x03]:
+            raise PluginError(f"Unsupported Domain Account F revision encountered: {f.revision}")
+
+        if fk.revision not in [0x01, 0x02]:
+            raise PluginError(f"Unsupported SAM Key Data revision encountered: {fk.revision}")
+
+        if fk.revision == 0x01:  # SAM key revision 0x01 involving RC4 (samsrv.dll: KEDecryptKeyWithRC4)
+            rc4_key = MD5.new(fk.salt + aqwerty + syskey + anum).digest()
+            samkey_data = ARC4.new(rc4_key).encrypt(fk.key + fk.checksum)
+            samkey = samkey_data[:16]
+            checksum = samkey_data[16:]
+
+            if checksum != MD5.new(samkey + anum + samkey + aqwerty).digest():
+                raise PluginError("SAM key checksum validation failed!")
+            return samkey
+
+        else:  # SAM key revision 0x02 involving AES  (samsrv.dll: KEDecryptKeyWithAES)
+            fk = c_sam.SAM_KEY_AES(f_key)
+            key_data = f_key[len(c_sam.SAM_KEY_AES) : len(c_sam.SAM_KEY_AES) + fk.data_len]
+            checksum_data = f_key[
+                len(c_sam.SAM_KEY_AES) + fk.data_len : len(c_sam.SAM_KEY_AES) + fk.data_len + fk.checksum_len
+            ]
+            samkey = SamPlugin.perform_aes(key=syskey, value=key_data, iv=fk.salt)[:16]
+            checksum = SamPlugin.perform_aes(key=syskey, value=checksum_data, iv=fk.salt)[:32]
+
+            if checksum != SHA256.new(samkey).digest():
+                raise PluginError("SAM key checksum validation failed!")
+            return samkey
+
+    @staticmethod
+    def expand_des_key(key: bytes) -> bytes:
+        # Expand the key from a 7-byte password key into an 8-byte DES key
+        key = bytearray(key[:7]).ljust(7, b"\x00")
+        s = bytearray(
+            [
+                ((key[0] >> 1) & 0x7F) << 1,
+                ((key[0] & 0x01) << 6 | ((key[1] >> 2) & 0x3F)) << 1,
+                ((key[1] & 0x03) << 5 | ((key[2] >> 3) & 0x1F)) << 1,
+                ((key[2] & 0x07) << 4 | ((key[3] >> 4) & 0x0F)) << 1,
+                ((key[3] & 0x0F) << 3 | ((key[4] >> 5) & 0x07)) << 1,
+                ((key[4] & 0x1F) << 2 | ((key[5] >> 6) & 0x03)) << 1,
+                ((key[5] & 0x3F) << 1 | ((key[6] >> 7) & 0x01)) << 1,
+                (key[6] & 0x7F) << 1,
+            ]
+        )
+        return bytes(s)
+
+    @staticmethod
+    def rid_to_key(rid: int) -> Tuple[bytes, bytes]:
```

Comment:
Could you move this to the global scope and out of the class?

Reply:
```suggestion
    def rid_to_key(rid: int) -> tuple[bytes, bytes]:
```
Also small suggestion to go with that :)